src/index.ts

```diff
@@ -1,21 +1,68 @@
-import { useSelector, useDispatch } from 'react-redux';
+import React, { createContext, useContext, useState } from 'react';
+import { useSelector, useDispatch, shallowEqual } from 'react-redux';
 import { useEffect } from 'react';
 import { Action } from 'redux';
+import {
+  ReactUseFetchWithReduxContext,
+  ReactUseFetchWithReduxProvider,
+} from './Provider';
+import hasCacheTimedOut from './utils/hasCacheTimedOut';
 
 function useFetchWithRedux<State, Selected>(
   getDataStart: () => Action,
   selector: (state: State) => Selected,
+  timeTillCacheInvalidateLocal?: number,
 ) {
   const dispatch = useDispatch();
-  const selected = useSelector(selector);
+  const selected = useSelector(selector, shallowEqual);
+  const [
+    cacheTimeouts,
+    setCacheTimeouts,
+    timeTillCacheInvalidateGlobal,
+  ] = useContext(ReactUseFetchWithReduxContext);
+
+  const cacheIndex = getDataStart().type;
+  const isCacheSet = Object.keys(cacheTimeouts).includes(cacheIndex);
+
+  const cacheInvalidationTime =
+    timeTillCacheInvalidateLocal || timeTillCacheInvalidateGlobal || null;
 
   useEffect(() => {
-    if (selected === null) {
+    if (!isCacheSet) {
       dispatch(getDataStart());
+      if (cacheInvalidationTime) {
+        // @ts-ignore
+        setCacheTimeouts({
+          ...cacheTimeouts,
+          [cacheIndex]: {
+            cacheInvalidationTime,
+            cacheSet: Date.now(),
+          },
+        });
+      }
     }
-  }, [dispatch]);
+  }, [
```

Comment:
Whoa that's a lot of hook dependencies. Do all of them need to be here? This will affect performance if we need to list them all.

Reply:
the linter puts them there :(
I had a brief look into refactoring this to work around it but I'm not sure it will be possible.